### PR TITLE
Script Loading Strategies: Use hash map in has_only_delayed_dependents

### PR DESCRIPTION
**Pull request by westonruter**
Amends https://github.com/WordPress/wordpress-develop/pull/4391

Addresses https://github.com/WordPress/wordpress-develop/pull/4391#discussion_r1226045554 from @mukeshpanchal27

### CODE REVIEW

File: tests/phpunit/tests/dependencies/scripts.php [hunk 1]
```diff
@@ -276,6 +280,128 @@ public function test_delayed_dependent_with_blocking_dependency_not_enqueued( $s
...
+					wp_enqueue_script( 'foo', 'https://example.com/foo.js', array(), null, array( 'strategy' => 'defer' ) );
+					return 'foo';
+				},
+				'async_only' => false,
+				'expected'   => true,
+			),
+			'one_delayed_dependent'                => array(
+				'set_up'     => static function () {
+					wp_enqueue_script( 'foo', 'https://example.com/foo.js', array(), null, array( 'strategy' => 'defer' ) );
+					wp_enqueue_script( 'bar', 'https://example.com/bar.js', array( 'bar' ), null, array( 'strategy' => 'defer' ) );
```
Comment (joemcgill) [inline tests/phpunit/tests/dependencies/scripts.php]: Is this meant to depend on foo?
```suggestion
					wp_enqueue_script( 'bar', 'https://example.com/bar.js', array( 'foo' ), null, array( 'strategy' => 'defer' ) );
```

File: tests/phpunit/tests/dependencies/scripts.php [hunk 2]
```diff
@@ -276,6 +280,128 @@ public function test_delayed_dependent_with_blocking_dependency_not_enqueued( $s
...
+					wp_enqueue_script( 'bar', 'https://example.com/bar.js', array( 'foo' ), null, array( 'strategy' => 'defer' ) );
+					wp_enqueue_script( 'baz', 'https://example.com/bar.js', array( 'bar' ), null, array( 'strategy' => 'defer' ) );
+					return 'foo';
+				},
+				'async_only' => false,
+				'expected'   => true,
+			),
+			'async_only_with_defer_dependency'     => array(
+				'set_up'     => static function () {
+					wp_enqueue_script( 'foo', 'https://example.com/foo.js', array( 'baz' ), null, array( 'strategy' => 'async' ) );
```
Comment (joemcgill) [inline tests/phpunit/tests/dependencies/scripts.php]: Is the `baz` dependency supposed to be in this test?
```suggestion
					wp_enqueue_script( 'foo', 'https://example.com/foo.js', array(), null, array( 'strategy' => 'async' ) );
```

File: tests/phpunit/tests/dependencies/scripts.php [hunk 3]
```diff
@@ -276,6 +280,128 @@ public function test_delayed_dependent_with_blocking_dependency_not_enqueued( $s
...
+					wp_enqueue_script( 'bar', 'https://example.com/bar.js', array( 'foo' ), null, array( 'strategy' => 'defer' ) );
+					return 'foo';
+				},
+				'async_only' => true,
+				'expected'   => false,
+			),
+			'async_not_only_with_defer_dependency' => array(
+				'set_up'     => static function () {
+					wp_enqueue_script( 'foo', 'https://example.com/foo.js', array( 'baz' ), null, array( 'strategy' => 'async' ) );
```
Comment (joemcgill) [inline tests/phpunit/tests/dependencies/scripts.php]: Same here?
```suggestion
					wp_enqueue_script( 'foo', 'https://example.com/foo.js', array(), null, array( 'strategy' => 'async' ) );
```

**Review by joemcgill — APPROVED** [review]
I've got a couple questions about the unit tests, but nothing blocking. This looks good to me and should help with performance when we're checking a large number of dependencies. It would be interesting to see some data about how much this change improves the performance of this algorithm.

Thanks @westonruter, and really good suggestion @mukeshpanchal27!

File: tests/phpunit/tests/dependencies/scripts.php (hunk 3 above)
Comment (westonruter) [reply tests/phpunit/tests/dependencies/scripts.php]: Fixed in f3e70bfbb92be1a940a8e9481e9048f8b2b5dba5

**Pull request merged**